configuration/postgres/postgres.go

```diff
@@ -0,0 +1,401 @@
+	InfoStartInit                = "initializing postgreSQL configuration store"
+	ErrorMissingConnectionString = "missing postgreSQL connection string"
+	ErrorAlreadyInitialized      = "PostgreSQL configuration store already initialized"
+	ErrorMissinMaxTimeout        = "missing PostgreSQL maxTimeout setting in configuration"
+	QueryTableExists             = "SELECT EXISTS (SELECT FROM pg_tables where tablename = $1)"
+	maxIdentifierLength          = 64 // https://www.postgresql.org/docs/current/limits.html
+	ErrorTooLongFieldLength      = "field name is too long"
+)
+
+func NewPostgresConfigurationStore(logger logger.Logger) configuration.Store {
+	logger.Debug("Instantiating PostgreSQL configuration store")
+	return &ConfigurationStore{
+		logger: logger,
+	}
+}
+
+func (p *ConfigurationStore) Init(metadata configuration.Metadata) error {
+	p.logger.Debug(InfoStartInit)
+
+	if p.client != nil {
+		return fmt.Errorf(ErrorAlreadyInitialized)
+	}
+	if m, err := parseMetadata(metadata); err != nil {
+		p.logger.Error(err)
+		return err
+	} else {
+		p.metadata = m
+	}
+	p.ActiveSubscriptions = make(map[string]*subscription)
+	ctx, cancel := context.WithTimeout(context.Background(), p.metadata.maxIdleTime)
+	defer cancel()
+	client, err := Connect(ctx, p.metadata.connectionString, p.metadata.maxIdleTime)
+	if err != nil {
+		return err
+	}
+	p.client = client
+	pingErr := p.client.Ping(ctx)
+	if pingErr != nil {
+		return pingErr
+	}
+	// check if table exists
+	exists := false
+	err = p.client.QueryRow(ctx, QueryTableExists, p.metadata.configTable).Scan(&exists)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *ConfigurationStore) Get(ctx context.Context, req *configuration.GetRequest) (*configuration.GetResponse, error) {
+	if err := validateInput(req.Keys); err != nil {
+		p.logger.Error(err)
+		return nil, err
+	}
+	query, params, err := buildQuery(req, p.metadata.configTable)
+	if err != nil {
+		p.logger.Error(err)
+		return nil, err
+	}
+	rows, err := p.client.Query(ctx, query, params...)
+	if err != nil {
+		// If no rows exist, return an empty response, otherwise return the error.
+		if err == sql.ErrNoRows {
+			return &configuration.GetResponse{}, nil
+		}
+		return nil, err
+	}
+	items := make(map[string]*configuration.Item)
+	for i := 0; rows.Next(); i++ {
+		var item configuration.Item
+		var key string
+		var metadata []byte
+		v := make(map[string]string)
+
+		if err := rows.Scan(&key, &item.Value, &item.Version, &metadata); err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(metadata, &v); err != nil {
+			return nil, err
+		}
+		item.Metadata = v
+		if item.Value != "" {
+			items[key] = &item
+		}
+	}
+	return &configuration.GetResponse{
+		Items: items,
+	}, nil
+}
+
+func (p *ConfigurationStore) Subscribe(ctx context.Context, req *configuration.SubscribeRequest, handler configuration.UpdateHandler) (string, error) {
+	var triggers []string
```

Comment:
Should this method have a lock inside to prevent multiple concurrent operations? I see that there are things that are not fully safe for concurrent use, such as `isSubscriptionActive`.

Reply:
https://github.com/dapr/components-contrib/blob/4f64eefae64507ca5cfd5f11044283816b126d4a/configuration/postgres/postgres.go#L153